pkg/alertmanager/alert_observer.go

```diff
@@ -0,0 +1,286 @@
+	if alerts == nil || o.limiter == nil || o.limiter.Level() <= 0 {
+		return
+	}
+	// The general idea of having levels in the limiter is to control the volume of logs AM is producing. The observer
+	// logs many types of events and some events are more important than others. By configuring the level we can
+	// continue to have observability on the alerts at lower granularity if we see that the volume of logs is getting
+	// too expensive.
+	// What is log per level is as follows:
+	//* 1
+	//	* Alert is rejected because of validation error
+	//	* Alert joins an aggregation group
+	//	* Alert is muted
+	//	* Aggregation Group notification Sent / Failed
+	//* 2
+	//	* Alert is rejected because of validation error
+	//	* Alert joins an aggregation group
+	//	* Alert is muted
+	//	* Aggregation Group pipeline start
+	//	* Aggregation Group notification Sent / Failed
+	//* 3
+	//	* Alert is rejected because of validation error
+	//	* Alert joins an aggregation group
+	//	* Alert is muted
+	//	* Aggregation Group pipeline start
+	//	* Aggregation Group passed a stage in the pipeline
+	//	* Aggregation Group notification Sent / Failed
+	//* 4
+	//	* Alert is rejected because of validation error
+	//	* Alert joins an aggregation group
+	//	* Alert is muted
+	//	* Aggregation group pipeline start
+	//	* Aggregation Group passed a stage in the pipeline
+	//	* Alert in aggregation group is Sent / Failed
+	//* 5
+	//	* Alert is rejected because of validation error
+	//	* Alert is received
+	//	* Alert joins an aggregation group
+	//	* Alert is muted
+	//	* Alert in aggregation group pipeline start
+	//	* Aggregation Group passed a stage in the pipeline
+	//	* Alert in aggregation group is Sent / Failed
+
+	switch event {
+	case alertobserver.EventAlertReceived:
+		o.Received(alerts)
+	case alertobserver.EventAlertRejected:
+		o.Rejected(alerts, meta)
+	case alertobserver.EventAlertAddedToAggrGroup:
+		o.AddedAggrGroup(alerts, meta)
+	case alertobserver.EventAlertFailedAddToAggrGroup:
+		o.FailedAddToAggrGroup(alerts, meta)
+	case alertobserver.EventAlertPipelineStart:
+		o.PipelineStart(alerts, meta)
+	case alertobserver.EventAlertPipelinePassStage:
+		o.PipelinePassStage(alerts, meta)
+	case alertobserver.EventAlertSent:
+		o.Sent(alerts, meta)
+	case alertobserver.EventAlertSendFailed:
+		o.SendFailed(alerts, meta)
+	case alertobserver.EventAlertMuted:
+		o.Muted(alerts, meta)
+	}
+}
+
+func (o *LogAlertLifeCycleObserver) Received(alerts []*types.Alert) {
+	if o.limiter.Level() < 5 {
+		return
+	}
+	for _, a := range alerts {
+		o.logWithAlert(a, true, "msg", "Received")
+	}
+}
+
+func (o *LogAlertLifeCycleObserver) Rejected(alerts []*types.Alert, meta alertobserver.AlertEventMeta) {
+	reason, ok := meta["msg"]
+	if !ok {
+		reason = "Unknown"
+	}
+	for _, a := range alerts {
+		o.logWithAlert(a, true, "msg", "Rejected", "reason", reason)
```

Comment:
do we need reject per log? maybe log rejected alerts in the same line better for less log volume?